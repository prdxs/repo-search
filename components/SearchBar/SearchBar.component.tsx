import React from 'react';
import clsx from 'clsx';
import styled from 'styled-components';
import Paper from '@material-ui/core/Paper';
import IconButton from '@material-ui/core/IconButton';
import InputBase from '@material-ui/core/InputBase';
import SearchIcon from '@material-ui/icons/Search';
import ClearIcon from '@material-ui/icons/Clear';

import { ISearchBarProps } from './SearchBar.typings';

const SearchBar: React.FC<ISearchBarProps> = ({
  className,
  style,
  value,
  onChange,
}) => {
  const hasValue = value !== undefined && value.length !== 0;

  return (
    <Paper className={clsx('SearchBar-root', className)} style={style}>
      <InputBase
        className="SearchBar-input"
        placeholder="Search users or repos"
        inputProps={{
          'aria-label': 'search users or repos',
          'data-testid': 'SearchBar-actualInput',
        }}
        value={value}
        onChange={onChange}
        data-testid="SearchBar-input"
      />
      {hasValue ? (
        <IconButton
          className=".SearchBar-iconButton"
          aria-label="clear"
          data-testid="SearchBar-clearButton"
        >
          <ClearIcon />
        </IconButton>
      ) : (
        <IconButton
          className=".SearchBar-iconButton"
          aria-label="search"
          data-testid="SearchBar-searchButton"
        >
          <SearchIcon />
        </IconButton>
      )}
    </Paper>
  );
};

const StyledSearchBar = styled(SearchBar)`
  &.SearchBar-root {
    padding: 2px 4px;
    display: flex;
    align-items: center;
    width: 400px;

    .SearchBar-input {
      margin-left: ${({ theme }) => theme.spacing(4)};
      flex: 1;
    }

    .SearchBar-divider {
      height: 28px;
      margin: 4px;
    }

    .SearchBar-iconButton {
      padding: 10px;
    }
  }
`;

export default StyledSearchBar;
