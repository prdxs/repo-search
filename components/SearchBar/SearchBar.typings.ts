import { IComponentProps } from '@/typings/common';

export interface ISearchBarProps extends IComponentProps {
  value: string;
  onChange(
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ): void;
}
