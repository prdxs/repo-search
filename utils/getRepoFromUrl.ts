const getRepoFromUrl = (url: string): { owner: string; repo: string } => {
  const parts = url.split('/');
  const len = parts.length;
  const owner = parts[len - 2];
  const repo = parts[len - 1];

  return { owner, repo };
};

export default getRepoFromUrl;
