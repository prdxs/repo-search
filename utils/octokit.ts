import { Octokit } from '@octokit/rest';

const octokit = new Octokit({
  baseUrl: 'https://api.github.com',
  auth: process.env.NEXT_PUBLIC_GH_TOKEN,
});

export default octokit;
