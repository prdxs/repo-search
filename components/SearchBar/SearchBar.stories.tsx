import React, { useState, useCallback } from 'react';
import { action } from '@storybook/addon-actions';

import StorybookWrapper from '@/components/StorybookWrapper';
import SearchBar from './SearchBar.component';

export default {
  title: 'SearchBar',
  component: SearchBar,
};

export const Controlled: React.FC = () => {
  const [value, setValue] = useState('');

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      const newValue = event.target.value;

      action('changed')(event);
      setValue(newValue);
    },
    []
  );

  const handleClear = useCallback(() => {
    action('cleared')();
    setValue('');
  }, []);

  return (
    <StorybookWrapper>
      <SearchBar value={value} onChange={handleChange} onClear={handleClear} />
    </StorybookWrapper>
  );
};
