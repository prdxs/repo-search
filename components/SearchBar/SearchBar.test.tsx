import React from 'react';
import { render, fireEvent } from '@/test/utils';
import SearchBar from '@/components/SearchBar';

describe('SearchBar', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<SearchBar value="" onChange={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain provided value', () => {
    const value = 'cat';
    const { getByDisplayValue } = render(
      <SearchBar value={value} onChange={jest.fn()} />
    );
    const input = getByDisplayValue(value);

    expect(input).toBeInTheDocument();
  });

  it("should render the search button when the input doesn't have a value", async () => {
    const { queryByTestId } = render(
      <SearchBar value="" onChange={jest.fn()} />
    );
    const searchButton = queryByTestId('SearchBar-searchButton');
    const clearButton = queryByTestId('SearchBar-clearButton');

    expect(searchButton).toBeInTheDocument();
    expect(clearButton).not.toBeInTheDocument();
  });

  it('should render the clear button when the input has a value', async () => {
    const value = 'dog';
    const { queryByTestId } = render(
      <SearchBar value={value} onChange={jest.fn()} />
    );
    const searchButton = queryByTestId('SearchBar-searchButton');
    const clearButton = queryByTestId('SearchBar-clearButton');

    expect(searchButton).not.toBeInTheDocument();
    expect(clearButton).toBeInTheDocument();
  });

  it('should call change handler when the value changes', async () => {
    const value = 'rat';
    const handleChange = jest.fn();
    const { getByTestId } = render(
      <SearchBar value="" onChange={handleChange} />
    );
    const input = getByTestId('SearchBar-actualInput');

    fireEvent.change(input, { target: { value } });

    expect(handleChange).toHaveBeenCalled();
  });
});
