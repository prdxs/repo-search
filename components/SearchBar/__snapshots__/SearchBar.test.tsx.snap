// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="padding: 2rem;"
  >
    <div
      class="MuiPaper-root SearchBar-root SearchBarcomponent__StyledSearchBar-sc-1icb8rc-0 fPthCZ MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiInputBase-root SearchBar-input"
        data-testid="SearchBar-input"
      >
        <input
          aria-label="search users or repos"
          class="MuiInputBase-input"
          data-testid="SearchBar-actualInput"
          placeholder="Search users or repos"
          type="text"
          value=""
        />
      </div>
      <button
        aria-label="search"
        class="MuiButtonBase-root MuiIconButton-root .SearchBar-iconButton"
        data-testid="SearchBar-searchButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
